Mentions

paulsonnentag opened the pull request:
WIP

paulsonnentag committed:
add search patterns

paulsonnentag committed:
add quill mention plugin

paulsonnentag committed:
externalize auto completion definition of quill plugin

paulsonnentag committed:
use text replacements instead of blots

paulsonnentag committed:
hide dropdown overlay if cursor is moved into text that matches autocompletion

paulsonnentag committed:
Add SOC autocomplete

paulsonnentag committed:
highlight docs with unread changes in content list with darker text color

pvh commented on src/components/content-types/ContentList.tsx:
```diff
@@ -214,14 +195,40 @@ export default function ContentList({ documentId }: ContentProps) {
+                  context="list"
+                  url={url}
+                  editable={contentUrlBeingEdited === url} // todo: this should be done in a more principled way
+                  onBlur={() => setContentUrlBeingEdited(undefined)}
```
So... let's talk about `onBlur`, probably in realtime with you me and Geoffrey. My instinct is that we should strongly avoid making the interface wider. There are a lot of incidental changes in the patch so I haven't internalized it all yet but I wonder how essential it is.

paulsonnentag committed:
detect mentions in chat messages

paulsonnentag committed:
detect mentions in text notes

paulsonnentag committed:
in changedDocsList only show documents with unseen mentions

paulsonnentag committed:
add per document notification settings

paulsonnentag committed:
show notification dot based on notification settings

paulsonnentag committed:
store notification settings per user

paulsonnentag committed:
fix text to show changed dot based on settings

paulsonnentag committed:
tweak style/layout

paulsonnentag committed:
restrict formats

paulsonnentag committed:
use workspace contacts for mention autocompletion

geoffreylitt commented on src/components/content-types/TextContent.tsx:
```diff
@@ -328,23 +371,38 @@ function TextInList(props: EditableContentProps) {
+  )
+}
+
+// TODO: this is not really checking if the user has been mentioned since the doc was last seen
```
Just to confirm, is the plan to defer this todo for now?

paulsonnentag replied:
I think we should wait until we get support for marks. Then, we can replace mentions with a custom object. That will make checking if you have unseen mentions trivial.

geoffreylitt commented on src/components/content-types/workspace/NotificationSetting.tsx:
```diff
@@ -0,0 +1,148 @@
+      }
+      placement="bottom"
+    >
+      <ListMenuSection title="Notify of changes">
```
How about clarify the copy like this:

<img width="219" alt="CleanShot 2022-12-13 at 17 47 03@2x" src="https://user-images.githubusercontent.com/934016/207461199-1554b279-c764-41ed-97fa-f067de6d8268.png">

paulsonnentag replied:
that's better

geoffreylitt reviewed (commented):
Generally looks good to me! Seems to work nicely. A few notes:

- Can you remove the context menu stuff so that this PR can be merged independently? (Unless I'm missing some reason why they need to be linked)
- It doesn't handle mentions of usernames that contain numbers. We should consider what valid characters are in usernames, but numbers seem reasonable to include?
- I think the design is a bit confusing. The notification settings only apply to the current doc, but the inbox is global. Actually the inbox is the odd one out here, because everything else is specific to this doc... Some small suggestions to tweak the design:
  - Use a settings icon for the notification settings, otherwise I'd expect to find notifications there
  - Clarify the copy (left a comment on the diff)
  - Move the inbox to the left side of the toolbar, that way it's closer to global things like search

paulsonnentag committed:
make list item bold if it has unseen changes

paulsonnentag committed:
tweak design

paulsonnentag committed:
allow any character in mentions except for spaces

paulsonnentag committed:
replace text input of ThreadContent with quill to support mentions

paulsonnentag committed:
add mark all as read button

paulsonnentag commented:
I've removed the context menu. Instead of changing the color of the text, I'm now changing the font weight to highlight documents with unread changes. That works fine with the text input state. 

- I've reordered the navbar. It's not great but at least it's more consistent
<img width="723" alt="Screen Shot 2022-12-14 at 16 00 23" src="https://user-images.githubusercontent.com/650540/207631029-055f4621-c1bc-4cc4-94be-ad236051c5ce.png">
I still prefer the bell icon over just a gear icon because you can see on what level you are subscribed to a document at a single glance. The orange background state is not ideal maybe a filled-out bell would be better. 

- I've reworded the title of the notification settings
- I made mentions more general so any character other than a spaces is allowed in a user name
- mentions autocompletion also works in threads now
   - the conditional text formatting doesn't, I could fix this but I think it's not worth it because we will move off quill eventually 
      anyways
- I've added a mark everything as read button

geoffreylitt commented:
@paulsonnentag this is really great! Just tried it out locally and all seems to work well. Having mentions in threads is huge. I think it's good to merge. 

Some notes that I think can be addressed post merge: (I could try addressing these this week)

- As you noted, the top left navbar is super cluttered  🙃 My proposal is to make the sidebar a permanent part of the system UI and move "global" stuff into the sidebar, like Notion. I think this solves a lot of problems; imo we need a more permanent frame to provide a stable sense of place

![CleanShot 2022-12-14 at 10 26 00@2x](https://user-images.githubusercontent.com/934016/207637299-22f6f3b7-7ebb-4c7d-a4f0-df27f05c9db7.png)

- Using font weight for unread docs is OK, but it doesn't feel quite right to me yet... can play with some alternatives
- It would be nice to have more context in the Inbox list for what exactly changed / how you were mentioned

paulsonnentag commented:
Agree with your comments. Unless @pvh has some more notes I would merge it as it is and create a new pull request for further refinements.

pvh commented on src/main.tsx:
```diff
@@ -69,26 +89,65 @@ async function introduceWorkers(sharedWorker: SharedWorker) {
   })
+}
+
+async function setupRepoWithoutSharedWorker() {
```
i don't love this part of the patch and it's not well-integrated with the rest of the system. would you split this into its own PR please?

some notes on this stuff:
- if there's no serviceworker we want to have a sharePolicy with the sync-server (otherwise new documents won't be pushed)
- we can do a better job of code reuse i suspect between the shared worker and here by making a separate file and putting this code into it
- does the serviceworker work when you do this? it's gonna block the main UI really badly.
- maybe we should just use a webworker in this situation?
- what breaks if you have two tabs writing at the same time. i don't think we do this right now but i'm pretty sure there are terrible data-corrupting race conditions lurking here.

paulsonnentag replied:
Sorry, that commit isn't supposed to be in here. I've started working on making blutack work without shared workers on a general fixes branch. I then rebased it onto this branch and later removed the shared workers logic but forgot to remove it here.

pvh commented on src/components/content-types/TextContent.tsx:
```diff
@@ -31,8 +31,15 @@ import {
+  evalSearchFor,
+  Match,
+  MENTION,
+} from "../pushpin-code/Searches"
```
this is an aside but we should rename pushpin-code to a more intellgible name. that was just where i put code that i pulled in from the pushpin project file-by-file as i discovered the minimum required set of stuff.

paulsonnentag replied:
I agree we should have a general cleanup patch after all the pull requests are merged. There are also a lot of unused imports and things like that

pvh commented:
Okay, I spent some time testing this patch and basically it doesn't seem to work for me. I wasn't able to make notifications work and on top of that, for me many documents simply stop loading entirely with this patch enabled. 

I'd like to spend some more time working on it but I have to go get my kids and make dinner now.

paulsonnentag commented:
> Okay, I spent some time testing this patch and basically it doesn't seem to work for me. I wasn't able to make notifications work and on top of that, for me many documents simply stop loading entirely with this patch enabled.
> 
> I'd like to spend some more time working on it but I have to go get my kids and make dinner now.

What do you mean specifically by notifications are not working? Do the documents that don't load have any shared characteristics?

pvh commented:
> What do you mean specifically by notifications are not working? Do the documents that don't load have any shared characteristics?

What I expected was that if I @mention someone in a comment it would show up as a changed document somewhere in the UI. I couldn't get that to happen in my testing.

paulsonnentag committed:
add changes line

paulsonnentag committed:
Add day line and tweak design of unread line

paulsonnentag committed:
fix missing react key

paulsonnentag committed:
Merge branch 'main' into mentions

paulsonnentag committed:
extend hasBadge lense with mention information

paulsonnentag committed:
merge 'main' into 'mentions'

paulsonnentag committed:
Merge branch 'main' into mentions

paulsonnentag committed:
add camera controls
